lib/open_food_network/sales_tax_report.rb

```diff
@@ -24,19 +31,39 @@ def orders
     end
 
     def table
-      orders.map do |order|
-        totals = totals_of order.line_items
-        shipping_cost = shipping_cost_for order
-
-        [order.number, order.created_at, totals[:items], totals[:items_total],
-         totals[:taxable_total], totals[:sales_tax], shipping_cost, order.shipping_tax, order.enterprise_fee_tax, order.total_tax,
-         order.bill_address.full_name, order.distributor.andand.name]
+      case params[:report_type]
+      when "tax_types"
+        orders.map do |order|
+          totals = totals_of order.line_items
+          shipping_cost = shipping_cost_for order
+
+          [order.number, order.created_at, totals[:items], totals[:items_total],
+           totals[:taxable_total], totals[:sales_tax], shipping_cost, order.shipping_tax, order.enterprise_fee_tax, order.total_tax,
+           order.bill_address.full_name, order.distributor.andand.name]
+        end
+      else
+        orders.map do |order|
+          [order.number, order.total - order.total_tax] +
+            relevant_rates.map { |rate| order.tax_adjustment_totals.fetch(rate, 0) } +
+            [order.total_tax, order.display_total]
+        end
       end
+
     end
 
 
     private
 
+    def relevant_rates
+      queries = [ search.result.joins(:line_items => {:adjustments => :tax_rate}).select('spree_tax_rates.*').uniq,
+                  search.result.joins(:adjustments => :tax_rate).select('spree_tax_rates.*').uniq ]
+      queries.map do |query|
+        ActiveRecord::Base.connection.select_all(query)
+      end.sum.map do
```

Comment:
It's already kind of in its separate method. If you mean have it in a model/decorator I can't as it builds upon the report search. I agree it's not the prettiest function but it was the only way I found to get relevant rates without making lots of SQL request (actually it perfoms only 2, one for tax adjustments on the order and one for tax adjustments on line_items [1])
[1] for instance : 
```
SELECT DISTINCT spree_tax_rates.* FROM "spree_orders" INNER JOIN "spree_line_items" ON "spree_line_items"."order_id" = "spree_orders"."id" INNER JOIN "spree_adjustments" ON "spree_adjustments"."adjustable_id" = "spree_line_items"."id" AND "spree_adjustments"."adjustable_type" = 'Spree::LineItem' INNER JOIN "spree_tax_rates" ON "spree_tax_rates"."id" = "spree_adjustments"."originator_id" AND spree_adjustments.originator_type = 'Spree::TaxRate' WHERE "spree_orders"."id" IN (4, 3, 9, 10, 7) AND (completed_at IS NOT NULL) AND (state != 'canceled') AND ("spree_orders"."completed_at" > '2015-12-31 23:00:00.000000')
SELECT DISTINCT spree_tax_rates.* FROM "spree_orders" INNER JOIN "spree_adjustments" ON "spree_adjustments"."adjustable_id" = "spree_orders"."id" AND "spree_adjustments"."adjustable_type" = 'Spree::Order' INNER JOIN "spree_tax_rates" ON "spree_tax_rates"."id" = "spree_adjustments"."originator_id" AND spree_adjustments.originator_type = 'Spree::TaxRate' WHERE "spree_orders"."id" IN (4, 3, 9, 10, 7) AND (completed_at IS NOT NULL) AND (state != 'canceled') AND ("spree_orders"."completed_at" > '2015-12-31 23:00:00.000000')
```